File: ackermann_ekf/tools/ekf_functions.py
#!/usr/bin/python3

import inspect
import sys

try:
    from sympy import *
except:
    print('Please install sympy first using\n\n    '
          'pip3 install sympy\n', file=sys.stderr)
    sys.exit(1)


def rpy2mat(roll, pitch, yaw):
    cr, sr = cos(roll), sin(roll)
    cp, sp = cos(pitch), sin(pitch)
    cy, sy = cos(yaw), sin(yaw)

    return Matrix([[cy*cp, cy*sp*sr - cr*sy, sy*sr + cy*cr*sp],
                   [cp*sy, cy*cr + sy*sp*sr, cr*sy*sp - cy*sr],
                   [-sp,             cp*sr,            cp*cr]])


def mat2rpy(M):
    return atan2(M[2, 1], M[2, 2]), -asin(M[2, 0]), atan2(M[1, 0], M[0, 0])


def sympy2defstr(args, expr, name):
    return inspect.getsource(lambdify(args, expr, 'math')) \
        .replace('_lambdifygenerated', name) \
        .replace('ImmutableDenseMatrix', 'array')


var('X Y Z v a ROLL PITCH YAW droll_dx dpitch_dx dyaw_dx')
var('sensor_x sensor_y sensor_z')
var('dt')

v_vec = Matrix([v, 0, 0])
a_vec = Matrix([a, 0, 0])
XYZ = Matrix([X, Y, Z])
sensor_xyz = Matrix([sensor_x, sensor_y, sensor_z])

print('Calculating M...')
M = rpy2mat(ROLL, PITCH, YAW)

print('Calculating M1...')
M1 = M*rpy2mat(droll_dx*v*dt, dpitch_dx*v*dt, dyaw_dx*v*dt)

print('Calculating dM_dt...')
dM_dt = ((M1 - M)/dt).limit(dt, 0)

print('Calculating RPY1...')
ROLL1, PITCH1, YAW1 = mat2rpy(M1)

print('Calculating XYZ1...')
XYZ1 = XYZ + M*v_vec*dt + (dM_dt*v_vec + M*a_vec)*dt**2/2

state = Matrix([XYZ,
                v,
                a,
                ROLL,
                PITCH,
                YAW,
                droll_dx,
                dpitch_dx,
                dyaw_dx])

print('Calculating f...')
f = simplify(Matrix([series(XYZ1[0], dt, n=3).removeO(),
                     series(XYZ1[1], dt, n=3).removeO(),
                     series(XYZ1[2], dt, n=3).removeO(),
                     series(v + a*dt, dt, n=2).removeO(),
                     series(a, dt, n=1).removeO(),
                     series(ROLL1, dt, n=2).removeO(),
                     series(PITCH1, dt, n=2).removeO(),
                     series(YAW1, dt, n=2).removeO(),
                     series(droll_dx, dt, n=1).removeO(),
                     series(dpitch_dx, dt, n=1).removeO(),
                     series(dyaw_dx, dt, n=1).removeO()]))

print('Calculating F...')
F = simplify(f.jacobian(state))

omega = Matrix([droll_dx*v, dpitch_dx*v, dyaw_dx*v])
domega_dt = Matrix([droll_dx*a, dpitch_dx*a, dyaw_dx*a])

print('Calculating h...')
h = simplify(Matrix([XYZ + M*sensor_xyz,
                     v_vec + omega.cross(sensor_xyz),
                     a_vec + 2*omega.cross(v_vec)
                     + domega_dt.cross(sensor_xyz)
                     + omega.cross(omega.cross(sensor_xyz)),
                     ROLL,
                     PITCH,
                     YAW,
                     omega]))

print('Calculating H...')
H = simplify(h.jacobian(state))


print('Done!\n\n\n')
print(f"""
# =============================================================================
# | This document was autogenerated by python from                            |
# | {'{:<74}'.format(__file__)}|
# | EDITING THIS FILE BY HAND IS NOT RECOMMENDED                              |
# =============================================================================

from numpy import array
from math import *

{sympy2defstr([list(state)] + [dt], list(f), 'f')}

{sympy2defstr([list(state)] + [dt], F.tolist(), 'F')}

{sympy2defstr([list(state)] + [list(sensor_xyz)], list(h), 'h')}

{sympy2defstr([list(state)] + [list(sensor_xyz)], H.tolist(), 'H')}
""")

File: ackermann_ekf_optimized/matlab/ekf_functions.m
clc

syms X Y Z speed accel Roll Pitch Yaw droll_dx dpitch_dx dyaw_dx real
syms dx_dt dy_dt dz_dt d2x_dt2 d2y_dt2 d2z_dt2 droll_dt dpitch_dt dyaw_dt real
syms sensor_x sensor_y sensor_z real
syms dt real

assume(Roll > -pi & Roll < pi)
assume(Pitch > -pi/2 & Pitch < pi/2)
assume(Yaw > -pi & Yaw < pi)

speed_vec = [speed; 0; 0];
accel_vec = [accel; 0; 0];
omega = [droll_dx; dpitch_dx; dyaw_dx]*speed;
domega_dt = [droll_dx; dpitch_dx; dyaw_dx]*accel;
sensor_position = [sensor_x; sensor_y; sensor_z];

M = rpy2mat(Roll, Pitch, Yaw);
M1 = M*rpy2mat(omega(1)*dt, omega(2)*dt, omega(3)*dt);
dM_dt = limit((M1 - M)/dt, dt, 0);

[Roll1, Pitch1, Yaw1] = mat2rpy(M1);

XYZ = [X; Y; Z];

state = [XYZ;
         speed;
         accel;
         [Roll; Pitch; Yaw]; 
         [droll_dx; dpitch_dx; dyaw_dx]];

observables = [XYZ;
               [dx_dt; dy_dt; dz_dt];
               [d2x_dt2; d2y_dt2; d2z_dt2];
               [Roll; Pitch; Yaw]; 
               [droll_dt; dpitch_dt; dyaw_dt]];

f_complete = [XYZ + M*speed_vec*dt + (dM_dt*speed_vec + M*accel_vec)*dt^2/2;
              speed + accel*dt;
              accel;
              Roll1;
              Pitch1;
              Yaw1;
              droll_dx;
              dpitch_dx;
              dyaw_dx];

f = simplify([taylor2(state(1:3), f_complete(1:3), dt);
              taylor1(state(4), f_complete(4), dt);
              taylor0(state(5), f_complete(5), dt);
              taylor1(state(6:8), f_complete(6:8), dt);
              taylor0(state(9:11), f_complete(9:11), dt)]);

F = simplify(jacobian(f, state));

h = simplify([XYZ + M*sensor_position;
              speed_vec + cross(omega, sensor_position);
              accel_vec + 2*cross(omega, speed_vec) + cross(domega_dt, sensor_position) + cross(omega, cross(omega, sensor_position));
              Roll;
              Pitch;
              Yaw;
              omega]);

H = simplify(jacobian(h, state));







